### PR TITLE
Added visual shader tut

### PR DESCRIPTION
**Pull request by ZX-WT**
Here's something I find missing in the docs, this is an short introduction to the visual shader editor.
It is mainly based on this article https://godotengine.org/article/major-update-for-visual-shader-in-godot-3-2

It is almost done, I just need to fix some of the text and add some images

### CODE REVIEW

File: tutorials/shading/Visual_shader.rst [hunk 1]
```diff
@@ -0,0 +1,84 @@
+.. _doc_Visual_shader:
```
Comment (akien-mga) [inline tutorials/shading/Visual_shader.rst]: Note that the file name and this reference should both be lowercase.

**Comment by ZX-WT** [comment]
This is it. This is a rather brief introduction, I think someone else could add more details. Anyway, please proof read. @clayjohn

File: tutorials/shading/visual_shader.rst [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+- If you right-click on the graph, this menu will be called at the cursor position and the created node, in that case, will also be placed under that position; otherwise, it will be created at the graph's center.
+- It can be resized horizontally and vertically allowing more content to be shown. Size transform and tree content position are saved between the calls, so if you suddenly closed the popup you can easily restore its previous state.
+- The ``Expand All`` and ``Collapse All`` options in the drop-down option menu can be used to easily list the available nodes.
+
+When connecting any ``scalar`` output to ``vector`` input, all the 3 components of the vector will be the value of the scalar.
+
+When connecting any ``vector`` output to ``scalar`` input, the value of the scalar will be the dot product between the original vector and a Vector3(0.33,0.33,0.33).
+
+The popup has categorized the nodes, it is overwhelming to understand or use most of them, try to pick some of them, plug it into the output socket and observe what happens.
```
Comment (Chaosus) [inline tutorials/shading/visual_shader.rst]: I would like if you mention that it's possible to drag&drop nodes to the graph - somewhere at these paragraphs

File: tutorials/shading/index.rst [hunk 3]
```diff
@@ -8,6 +8,7 @@ Shading
    shading_reference/index
    your_first_shader/index
    shader_materials
+   visual_shader
```
Comment (clayjohn) [inline tutorials/shading/index.rst]: make it plural "visual_shaders"

File: tutorials/shading/visual_shader.rst [hunk 4]
```diff
@@ -0,0 +1,101 @@
+.. _doc_visual_shader:
+
+Visual Shader
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: Make plural and one word:

VisualShaders
=========

File: tutorials/shading/visual_shader.rst [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+=============
+
+Introduction
+------------
+
+VisualScript is an alternative for users that cannot code, Similarly, Visual Shader is the visual alternative for creating shaders.
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: "Just as VisualScript is an alternative for users that cannot code, VisualShaders are the visual alternative for creating shaders."

File: tutorials/shading/visual_shader.rst [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+.. note::
+    
+   If you do not understand what shader is , go read :ref:`doc_what_are_shaders`
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: "If you do not understand what shaders are, go read :ref:`doc_what_are_shaders`

File: tutorials/shading/visual_shader.rst [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+   
+Creating a VisualShader
+-----------------------
+
+Create a new ``ShaderMaterial`` in some object of your choice.
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: "VisualShaders can be created in any ShaderMaterial. To begin using VisualShaders, create a new ``ShaderMaterial`` in an object of your choice.

File: tutorials/shading/visual_shader.rst [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+The layout of the visual shader editor can be divided into two parts the upper bar and the graph.
+
+.. image:: img/visual_shader_editor2.png
+
+From left to right of the upper bar, the ``Add Node`` button allows you to add more nodes to the shader graph, to the right, is the shader mode : ``vertex``, ``fragment`` and ``light``. Then, it is the zooming level and the snapping and grid options. The last icon can show the resulting shader code that the graph creates.
+
+Using the Visual Shader Editor
+------------------------------
+
+By default, every new ``VisualShader`` will have an output node. The node connection usually ends at one of the output socket. A node is the basic unit to create your shader. To add a new node, click on the ``Add Node`` button on the upper left corner or right click on any empty space inside the Visual Shader Editor, a menu will pop up.
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: "Every node connection ends at one of the output node's sockets."

"...space inside the Visual Shader Editor, and a menu will pop up."

File: tutorials/shading/visual_shader.rst [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+This popup has the following properties:
+
+- If you right-click on the graph, this menu will be called at the cursor position and the created node, in that case, will also be placed under that position; otherwise, it will be created at the graph's center.
+- It can be resized horizontally and vertically allowing more content to be shown. Size transform and tree content position are saved between the calls, so if you suddenly closed the popup you can easily restore its previous state.
+- The ``Expand All`` and ``Collapse All`` options in the drop-down option menu can be used to easily list the available nodes.
+
+When connecting any ``scalar`` output to ``vector`` input, all the 3 components of the vector will be the value of the scalar.
+
+When connecting any ``vector`` output to ``scalar`` input, the value of the scalar will be the dot product between the original vector and a Vector3(0.33,0.33,0.33).
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: Vector3 should have backticks around it:
\`\`Vector3(0.33, 0.33, 0.33)\`\`

File: tutorials/shading/visual_shader.rst [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+By default, every new ``VisualShader`` will have an output node. The node connection usually ends at one of the output socket. A node is the basic unit to create your shader. To add a new node, click on the ``Add Node`` button on the upper left corner or right click on any empty space inside the Visual Shader Editor, a menu will pop up.
+
+.. image:: img/vs_popup.png
+
+This popup has the following properties:
+
+- If you right-click on the graph, this menu will be called at the cursor position and the created node, in that case, will also be placed under that position; otherwise, it will be created at the graph's center.
+- It can be resized horizontally and vertically allowing more content to be shown. Size transform and tree content position are saved between the calls, so if you suddenly closed the popup you can easily restore its previous state.
+- The ``Expand All`` and ``Collapse All`` options in the drop-down option menu can be used to easily list the available nodes.
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: Add one more bullet here to address @Chaosus's comment
e.g.
- You can also drag and drop nodes from the popup onto the graph

File: tutorials/shading/visual_shader.rst [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+When connecting any ``scalar`` output to ``vector`` input, all the 3 components of the vector will be the value of the scalar.
+
+When connecting any ``vector`` output to ``scalar`` input, the value of the scalar will be the dot product between the original vector and a Vector3(0.33,0.33,0.33).
+
+The popup has categorized the nodes, it is overwhelming to understand or use most of them, try to pick some of them, plug it into the output socket and observe what happens.
+
+.. note::
+    
+   Although ``VisualShader`` does not require coding, it shares the same logic with normal shaders, you can learn and convert shaders to visual shaders after understanding the logic behind the shader.
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: Although ``VisualShaders`` do not require coding, they share the same logic with normal shaders, you can learn and convert shaders to visual shaders after understanding the logic behind the shader.

File: tutorials/shading/visual_shader.rst [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+.. image:: img/vs_expression.gif
+
+The possibilities of this node are almost limitless – you can write complex procedures, and use all the power of Godot (text-based) shaders, such as loops, discard keyword, extended types, etc. For example:
+
+.. image:: img/vs_expression2.png
+
+Fresnel node
+++++++++++++
+
+The ``Fresnel`` node is designed to accept normal and view vectors and produces a scalar which is the saturated dot product between them. Additionally you can setup the inversion and the power of equation
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: add: 
The ``Fresnel`` node is great for adding a rim-like lighting effect to objects.

File: tutorials/shading/visual_shader.rst [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+.. image:: img/vs_fresnel.png
+
+Boolean node
+++++++++++++
+
+The ``Boolean`` node can be converted to Scalar or Vector to represent 0 or 1 and (0, 0, 0) or (1, 1, 1) respectively. This property can be used to enable or disable some effect parts by one click.
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: "with one click" instead of "by one click"

File: tutorials/shading/visual_shader.rst [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+.. image:: img/vs_if.png
+
+Switch node
++++++++++++
+
+The ``Switch`` node returns a vector if the boolean condition is true or false. ``Boolean`` was introduced above. If you convert a vector to a true boolean, all components of the vector should be above zero.
+
+.. image:: img/vs_switch.png
+
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: add:

.. note::
    The ``switch`` node is only available on the GLES3 backed. If you are targeting GLES2 devices, you cannot use ``switch`` statements.

File: tutorials/shading/visual_shader.rst [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+Derivative nodes
```
Comment (clayjohn) [inline tutorials/shading/visual_shader.rst]: In my opinion derivatives are a little advanced for an introductory tutorial. So if it were up to me I would cut this section out. However, it is your tutorial and I will leave the decision up to you. :)

File: tutorials/shading/visual_shader.rst (hunk 9 above)
Comment (akien-mga) [reply tutorials/shading/visual_shader.rst]: That's a convoluted way to express that the scalar will be the average of the three components :)

**Comment by akien-mga** [comment]
I fixed the merge conflict and proofread the tutorial, addressing the feedback given by @clayheaton and @Chaosus, and pushed the resulting commit as 1a977971 (the branch for this PR did not allow force pushes).

It's a great start on the tutorial, thanks! I reordered some things for more consistency, and we should probably describe more nodes in the future, but it's a good first version.

Future work can likely take inspiration/resources from https://godotengine.org/article/major-update-visual-shaders-godot-3-2 and https://godotengine.org/article/major-update-visual-shaders-godot-3-2-part-2 to describe further nodes.

**Comment by clayheaton** [comment]
@akien-mga I think you meant @clayjohn but this is cool to see!

**Comment by akien-mga** [comment]
Ah indeed, sorry for the mis-ping :D

**Comment by ZX-WT** [comment]
Ah, I am going to add more stuff after reading the new article for shader editor, will do it for another PR.

**Pull request closed without merging**